obs-outputs: Add dynamic variable bitrate

pkviet opened the pull request:
Suggestions link : [here](https://ideas.obsproject.com/posts/349/dynamic-bitrate-support)

**Updated (5/5/2018)**

Partly based on R1CH code from OBS Classic with some changes from Woodbyte fork.  
I'm removing the experimental tag since it's working quite well.    
The algo has seen a series of improvements, notably support for all h264 encoders supported by obs
(x264, enc-amf, nvenc, mac-vth264, quicksync, etc).    

**warning**
- For enc-amf, a flag signalling that dynamic bitrate must be added; I put this as a placeholder so that I don't forget to submit a PR to enc-amf submodule repo.  (see below update of 02/22)   
- for (old) nvenc, ffmpeg master git head is required since dynamic bitrate was added to nvenc ffmpeg only recently (05/05/18 ) through a patch I submitted. 
- new jim-nvenc supported. Works much better than previous nvenc. I routinely get zero frame drops.

**What this does:**
- for rtmp streams, decrease the video bitrate when available bandwidth is going down until all packets go through;
- increase the video bitrate once bandwidth conditions get better.

**Some screenshots**    
(1) enable dynamic bitrate in Simple output     

![capture5](https://user-images.githubusercontent.com/9283407/42186937-aa466ff2-7e4e-11e8-9b1c-15b75d82bdd4.PNG)

(2) Advanced settings for dynamic bitrate    

![capture4](https://user-images.githubusercontent.com/9283407/42186955-bb85ea68-7e4e-11e8-8a40-cc8ffe4942a4.PNG)

(3) Bitrate decrease from 2500 kbs due to congestion (downward red arrow)    

![capture](https://user-images.githubusercontent.com/9283407/42186999-e183e7a6-7e4e-11e8-806e-eef5f8b85db8.PNG)

(4) Bitrate stationary during congestion (not increasing yet because congestion hasn't cleared, arrow is now orange and not red as in this screenshot, see discussion with Dodgepong)    

![capture2](https://user-images.githubusercontent.com/9283407/42187061-0dab7718-7e4f-11e8-9e2d-1c7c035ddd55.PNG)

(5) Bitrate increasing due to congestion clearing (upward green arrow (red in this screenshot))

![capture3](https://user-images.githubusercontent.com/9283407/42187101-29945724-7e4f-11e8-8c38-0284e6bd035e.PNG)

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,12 +1095,77 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+	int initial_bitrate = stream->initial_bitrate;
+
+	// X264_NAL_HRD_CBR=2 incompatible with dynamic variable bitrate
+	if ( (strcmp(encoder_id, "obs_x264") == 0) && i_nal_hrd != 2 &&
```
Style: Do not put spaces between parentheses and their enclosed contents.

RytoEX commented on plugins/obs-x264/obs-x264.c:
```diff
@@ -205,6 +208,7 @@ static obs_properties_t *obs_x264_props(void *unused)
 			OBS_COMBO_TYPE_LIST, OBS_COMBO_FORMAT_STRING);
 	obs_property_list_add_string(list, TEXT_NONE, "");
 	add_strings(list, x264_tune_names);
+	obs_properties_add_bool(props, "dynamic_variable_bitrate", TEXT_DYN_BITRATE);
```
Style:  Break lines to avoid exceeding 80 columns.  Suggestion:

```
	obs_properties_add_bool(props, "dynamic_variable_bitrate",
			TEXT_DYN_BITRATE);
```

pkviet commented:
thanks again !

pkviet commented:
On Hold
======
Following a suggestion of Xaymar, I'll add enc-amf support. For now this plugin works only for x264.
I'll try to add also for nvenc and libmfx

pkviet commented:
PR Updated
=================
Support for enc-amf and QSV added.    
FFmpeg-nvenc is unfortunately incompatible with dynamical bitrate; the feature was not implemented.     
In terms of bitrate switching, the most efficient encoder is QSV ; enc-amf and x264  are on par, one better than the other in some conditions.     
The switching has been tested under hard-wall conditions: upload speed bounded (through a router),    
and on macOS by simulating various internet connexions using Network Link Conditioner (free XCode tool).

notr1ch commented on UI/data/locale/en-US.ini:
```diff
@@ -630,6 +631,7 @@ Basic.Settings.Output.Adv.Rescale="Rescale Output"
 Basic.Settings.Output.Adv.AudioTrack="Audio Track"
 Basic.Settings.Output.Adv.Streaming="Streaming"
 Basic.Settings.Output.Adv.ApplyServiceSettings="Enforce streaming service encoder settings"
+Basic.Settings.Output.Adv.DynamicalBitrate="Enable Dynamical Variable Bitrate (not supported for nvenc)"
```
I think "Dynamical Variable Bitrate" is a bit too hard for a user to understand what it does.

I would suggest "Auto adjust bitrate for network conditions" or "Auto manage bitrate if congestion detected" or something along those lines.

pkviet replied:
thanks R1CH, you're right, will change to what you suggest.
Dodgepong suggested to add  an indicator at the bottom when bitrate switching is active.
Wondering also if it would be useful to add some user control on the bitrate decrease and increase rate (that would be in Advanced mode obviously).

reboot commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1098,116 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
 {
+	obs_encoder_t *vencoder = obs_output_get_video_encoder(stream->output);
+	const char *encoder_id = obs_encoder_get_id(vencoder);
+	if (strcmp(encoder_id, "ffmpeg_nvenc") != 0) {
```
I think for future additions of more encoders it would be better to not hard code this exclusion but instead let the encoder specify if it supports dynamic bitrate changes or not.

pkviet replied:
that's a good point.   
That was the lazy option I took.    
It would require to add a flag to h264 encoders.   
Will do that as well as other improvements; I forgot to check videotoolbox for instance.

dodgepong commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -940,13 +940,46 @@ static void *connect_thread(void *data)
+		stream->dynamic_bitrate = bitrate;
+		bool isSimpleMode = obs_data_get_bool(settings, "IsSimpleMode1") ||
+				obs_data_get_bool(settings, "IsSimpleMode2");
+		blog(LOG_INFO, "the mode is %i \n", isSimpleMode);
```
What is the purpose of this log line?

Speaking of which, I see that there is a log entry when bitrate is adjusted, but is there an entry anywhere to indicate whether the user has enabled this feature at all?

pkviet replied:
oh i just forgot to remove the line which helped me during dev.   
It was to track the bool. It is used for UI . Initially I had coded only for x264 but if other h264 encoders are supported I needed the optio nto appear in Simple and Advanced mode as well.   
I'll log the activation of variable bitrate, good idea.

pkviet commented:
Update and to do
=================

- Added flags to h264 encoders ( through info.caps ) to signal dynamic bitrate support (cf @reboot ).    
- Added log when dynamic bitrate is enabled (cf @dodgepong )    
- Changed text displayed in UI for option (cf @notr1ch )
- Added an arrow in status bar to indicate whether bitrate switching is active and which way it is going (increasing bitrate, decreasing it  or leaving it) (cf @dodgepong on discord )    
- **TO DO:** enable finer user control on bitrate decrease or increase rate in UI (cf @jp9000 on discord who thought the decrease rate is quite steep; the right rate much depends on local network conditions so this would allow the user to adapt it a bit).    
I'll push additional commits when the last item is done.    
I'm separating the features in different commits for easier cherry-picking.

pkviet commented:
Update (02/22/2018)
===========
Applied the suggestions in the various reviews.

SuslikV commented:
1) I don't see the way to modify the icons for my custom theme (.qss). Is it even possible?

2) I have other thoughts for description text (the difference info).

`Basic.Settings.Output.DynamicBitrate="Auto manage bitrate decrease in case of network congestion"`

- (otherwise, I think that this option do miracle and auto-adjust my bitrate as magic Auto-Configuration Wizard. The encoder specific {NVENC} shouldn't be mentioned - you need to disable/hide that control - that's all.)

`Basic.Settings.Output.DynamicBitrate.Tooltip="If enabled and network congestion is detected, video bitrate\nwill be decreased by few % every second.\n\nOnce the congestion has cleared up, bitrate will increase every 5 sec,\nuntil target bitrate reached."`

- (not clear, should I leave this option unchecked or enable it? Exact numbers (%%) not needed because they are subject to change and not adjustable, or you should give detail description at what % will decrease/increase each encoder. I mean - "~about" is not an option here.)

`Basic.Settings.Output.Adv.DynamicBitrate="Auto manage bitrate decrease in case of network congestion"`

- (same as Basic.Settings.Output.DynamicBitrate)

`Basic.Settings.Output.Adv.DynamicBitrate.Tooltip="If enabled and network congestion is detected, video bitrate\nwill be decreased by few % every second.\n\nOnce the congestion has cleared up, bitrate will increase every 5 sec,\nuntil target bitrate reached."`

- (same as for Basic.Settings.Output.DynamicBitrate.Tooltip)

```
Basic.Settings.Output.Adv.Dyn.Down="Decrease bitrate, step (in %)"
Basic.Settings.Output.Adv.Dyn.Up="Increase bitrate, step (in %)"
```

- (of course, if rate is mentioned this should change only **rate** - because previous option says about **bitrate**, "stepping" - is option too, otherwise just leave it unchanged)

`Basic.Settings.Output.Adv.DynamicBitrate.Down.Tooltip="When congestion is detected, the video bitrate will be decreased\nevery second by the specified amount until congestion stops."`

- (I don't get what is "X %" here. If this variable then syntax is wrong. If not - leave only brief info here. Defaults may differ per encoder, if not - add "\n\nDefault value is 15%." at the end)

`Basic.Settings.Output.Adv.DynamicBitrate.Up.Tooltip="When congestion clears up, the video bitrate will be increased every\n5 seconds by the specified amount until target bitrate reached."`

- (I don't get what is "X %" here. If this is variable then syntax is wrong. If not - leave only brief info here. Defaults may differ per encoder, if not - add "\n\nDefault value is 10%." at the end)

3) I what to know (from interface) how much it can decrease the bitrate (read - quality of the stream) when enabled, until frame dropping will be forced. Maybe you should add this to tool tip of the option as "\n\nMinimum possible bitrate is half of the target bitrate, if congestion continues - frame drop will be forced."

pkviet commented:
@SuslikV 
1. check in UI/forms/images for the icons names. Then look at UI/data/themes/Rachni to see how you can override the icons.
2. Disabling the option for nvenc requires some UI work which I hesitate to implement at the moment because it will have to be reverted if ffmpeg-nvenc gets dynamic bitrate support (I'm thinking of submitting a patch to ffmepg eventually). But I'll look into it.
For the tooltips, texts, let me think about it and also in light of further reviews.
3. The dynamic bitrate mechanism is independent from the frame drop mechanism. It sets in much sooner than frame dropping so it can prevent frame dropping altogether if aggressive enough, though in general it won't prevent frame drops but will mitigate them. 
So it's not possible to quantify when frame dropping will start as a function of bitrate decrease.

SuslikV commented:
1. I don't get it. The Rachni alternates only general icons for widgets, not the one that is from the storage (obs.qrc). And few, that has custom string properties (like "themeID"). Can you explain more, how to access your images by Qt stylesheet?
2. As you wish.
3. Queue shortens faster? The bitrate has minimum? How can you explain to the end user why dynamic bitrate is needed then? If this feature not needed - no sense to implement it. For example, I know, if I have a lot of drops - I need to lower bitrate - it always helps.

pkviet commented:
1) look at the list of files in that PR: there are 3 png, dyn.png, dynup.png and dyndown.png ;
once merged they will be located in UI/forms/images ; if you check obs.qrc of this PR you will see there are three new entries.
3) if the user has say 1 Mbs of upload bandwidth but selects more, the  option will lower automatically the bitrate for him/her. If the upload bandwidth fluctuates , it will do also. And when bandwidth gets larger, the option will increase back the bitrate.   
If the bitrate is contained, it will of course decrease the number of frame drops. The option controls frame drops indirectly in this way: if bitrate decreases, buffer will not be as filled and therefore frame drops will be less frequent. 
When I say the option can only mitigate drops and not prevent them completely is because when the algo tries to increase the bitrate, it will test against congestion parameter (which is a measure of how filled is the buffer): if there is no congestion, the algo knows it can still increase the bitrate until congestion appears. Then as to frame drops occuring during that increase phase, it all depends on a combination of how fast network bandwidth fluctuates and how fast the algo adjusts bitrate. When decreasing bitrate, it is the same; suppose bandwidth is going down faster than the bitrate adjustment, the buffer will get filled and therefore there will be drops. But less than if the option had not been activated.    
What must be realized is that frame drops / dynamic bitrate are two different mechanisms to deal with fluctuating bandwidth. The first prioritizes the quality of the picture, with less frames; the second prioritizes the frames continuity to the expense of image quality.

SuslikV commented:
Oh, about the theme icons, I mean - custom (user) themes, for example: https://obsproject.com/forum/resources/categories/themes.10/
I thought, that it would be nice to have styled elements when you make changes to UI and implementing this new code in obs. And current implementation as hardcoded strings in the code - they are exist but you cannot adjust them without bit-hack.

The description is nice. Unfortunately it is huge. And thus tool tip with all this info is a bad idea. Also, user may don't know about the encoder's buffer, it fullness etc. Let's imagine that All what I (user) see is: auto option; current bitrate value; up/down arrows; frame drop or quality drops (from stream watchers feedback). And all this, is connected to user settings of the mentioned option (no matter how it works).
Now I thinking of naming the option: "Dynamic bitrate reduction" this is short and all other info can be moved into the tool tip.

pkviet commented:
1) Themes can override all icons, so these icons in my PR can also be overriden . But maybe I misunderstood you, sorry then. Might be simpler to discuss this on discord on dev channel.    
2) You're right , the end-user does not need too much info. Notice the controls for the algo only appear in Output > Advanced mode.    
I get your point about tooltips and the rest. I'm waiting for Jim's review to synthesize everything now. Thanks.

SuslikV commented:
In general I don't know how to select exact object. QLabel is selectable but all 3 states (up/down/static) is QPixmap picture that replaces each other. Thus I can give only one style to QLabel itself but not designate it states. After picture updates, the style needs to be updated in the code too (at least for this element), the simplest way is to set style with empty comment string. Also, to be able to address each state I need unique name for it (the simplest way is to add string to properties, for example "themeID"="up_bitrate") this name should change as soon as new state switched, but before style update. Thus it will be possible to use .qss:
```
QLabel[themeID="up_bitrate"] {
    my custom style...
}
```
Also, it is desirable to have unique and predefined "name" property for the widget (QLabel). Because potentially you can share same picture to other widgets and adjust only size of it for each widget type individually.

The chance to be implemented is small because I'm talking about the changes that were not accepted. It is different, but the root is the same. https://github.com/jp9000/obs-studio/pull/856

pkviet commented:
Update 03/14/18: fixed some UI bugs (never use Qt Creator on obs ui files !)

dodgepong commented:
Upon discussion in the discord, I actually think this feature should be enabled by default. For the average user, I think they would prefer to degrade quality and not drop frames rather than drop frames with static quality.

pkviet commented:
Update:
Fixed mem leaks.

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -393,6 +393,9 @@ void SimpleOutput::Update()
 	bool enforceBitrate = config_get_bool(main->Config(), "SimpleOutput",
 			"EnforceBitrate");
+	bool dynamicBitrate = config_get_bool(main->Config(), "SimpleOutput",
+		"DynamicBitrate");
```
Style:  Tab twice to indent a line continuation.

pkviet replied:
done

RytoEX replied:
Sorry.  I meant tab twice from the start of the previous line, like 394 and 395 above.

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1499,13 +1522,32 @@ bool AdvancedOutput::StartStreaming(obs_service_t *service)
+	bool dynamicBitrate = config_get_bool(main->Config(), "AdvOut",
+			"DynamicBitrate");
+	int dynamicBitrateDown = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateDown");
```
Style:  Tab twice to indent a line continuation.

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1499,13 +1522,32 @@ bool AdvancedOutput::StartStreaming(obs_service_t *service)
+	int dynamicBitrateDown = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateDown");
+	int dynamicBitrateUp = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateUp");
```
Style:  Tab twice to indent a line continuation.

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1499,13 +1522,32 @@ bool AdvancedOutput::StartStreaming(obs_service_t *service)
+	int dynamicBitrateUp = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateUp");
+	int dynamicBitrateThreshold = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateThreshold");
```
Style:  Tab twice to indent a line continuation.

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1499,13 +1522,32 @@ bool AdvancedOutput::StartStreaming(obs_service_t *service)
+	int dynamicBitrateThreshold = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateThreshold");
+	int dynamicBitrateRecoveryTime = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateRecoveryTime");
```
Style:  Tab twice to indent a line continuation.

RytoEX commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1499,13 +1522,32 @@ bool AdvancedOutput::StartStreaming(obs_service_t *service)
+	int dynamicBitrateRecoveryTime = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateRecoveryTime");
+	int dynamicBitrateDecreaseTime = config_get_int(main->Config(), "AdvOut",
+		"DynamicBitrateDecreaseTime");
```
Style:  Tab twice to indent a line continuation.

RytoEX commented on UI/window-basic-status-bar.cpp:
```diff
@@ -307,6 +333,28 @@ void OBSBasicStatusBar::UpdateDroppedFrames()
 
+void OBSBasicStatusBar::UpdateDynamicBitrateStatus() {
+	if (!streamOutput)
+		return;
```
Style:  Could use a blank newline between this line and the next line.

RytoEX commented on UI/window-basic-status-bar.cpp:
```diff
@@ -384,6 +432,12 @@ void OBSBasicStatusBar::UpdateStatusBar()
 
 	UpdateDroppedFrames();
+	variableBitrateEnabled = GetDynamicBitrateSetting();
+	if (variableBitrateEnabled) {
```
Style:  Curly braces are not needed for this `if-else` statement.

RytoEX commented on libobs/obs-output.c:
```diff
@@ -2129,6 +2129,14 @@ float obs_output_get_congestion(obs_output_t *output)
+int obs_output_get_bitrate_state(obs_output_t *output)
+{
+	if (!obs_output_valid(output, "obs_output_get_bitrate_state"))
+		return 0;
```
Style: Could use a blank newline between this line and the next line.

pkviet replied:
line added, dunno why github keeps this open

RytoEX commented on libobs/obs-output.c:
```diff
@@ -2129,6 +2129,14 @@ float obs_output_get_congestion(obs_output_t *output)
+	if (!obs_output_valid(output, "obs_output_get_bitrate_state"))
+		return 0;
+	int state = output->info.get_bitrate_state(output->context.data);
+	return state;
```
Why not just `return output->info.get_bitrate_state(output->context.data);`?

RytoEX commented on libobs/obs.h:
```diff
@@ -1565,6 +1565,8 @@ EXPORT void obs_output_output_caption_text1(obs_output_t *output,
 EXPORT float obs_output_get_congestion(obs_output_t *output);
 EXPORT int obs_output_get_connect_time_ms(obs_output_t *output);
 
+EXPORT int obs_output_get_bitrate_state(obs_output_t *output);
```
Maybe add a doc-comment above this function?

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -940,13 +940,61 @@ static void *connect_thread(void *data)
+	if (stream->switch_variable_bitrate) {
+		blog(LOG_INFO, "Dynamic bitrate ON: bitrate auto management"
+			" when network congestion is detected.\n");
+	}
```
Should be `} else {`

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+	obs_encoder_t *vencoder = obs_output_get_video_encoder(stream->output);
+	const char *encoder_id = obs_encoder_get_id(vencoder);
+	uint32_t caps = obs_get_encoder_caps(encoder_id);
+	/* */
```
Missing comment?

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1119,6 +1303,7 @@ static void rtmp_stream_data(void *data, struct encoder_packet *packet)
 		os_sem_post(stream->send_sem);
 	else
 		obs_encoder_packet_release(&new_packet);
+
```
Unnecessary newline.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+		 * So we choose to have lower quality instead of better with more drops.
+		 */
+		if (increase_br) {
+
```
Unnecessary newline.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+
+			current_bitrate = (int)((float)current_bitrate * (1.0 + increase_rate/100));
+
+			if (current_bitrate > initial_bitrate) {
```
Style:  This `if` statement does not need curly braces.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+			(cur_time_ms - last_adjustment_time) > polling_time;
+
+	if (i_nal_hrd != 2 && stream->switch_variable_bitrate) {
+
```
Unnecessary newline.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+		if (decrease_br) {
+			current_bitrate = (int)(current_bitrate / (1.0 + decrease_rate/100.0));
+
+			if (current_bitrate < (int)((float)initial_bitrate / 2.0)) {
```
Style:  This `if` statement does not need curly braces.

RytoEX commented on UI/window-basic-settings.cpp:
```diff
@@ -4183,6 +4222,7 @@ void OBSBasicSettings::SimpleRecordingEncoderChanged()
 	QString warning;
 	bool advanced = ui->simpleOutAdvanced->isChecked();
 	bool enforceBitrate = ui->simpleOutEnforce->isChecked() || !advanced;
+
```
Unnecessary newline.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+	if (i_nal_hrd != 2 && stream->switch_variable_bitrate) {
+
+		if (decrease_br) {
+			current_bitrate = (int)(current_bitrate / (1.0 + decrease_rate/100.0));
```
Style:  Math operators should always be surrounded by a single space on each side.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+		 */
+		if (increase_br) {
+
+			current_bitrate = (int)((float)current_bitrate * (1.0 + increase_rate/100));
```
Style: Math operators should always be surrounded by a single space on each side.

RytoEX commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1113,145 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+	const char *encoder_id = obs_encoder_get_id(vencoder);
+	uint32_t caps = obs_get_encoder_caps(encoder_id);
+	/* */
+	if (caps & OBS_ENCODER_VIDEO_DYN) {
```
Style:  This `if` statement does not need curly braces.

pkviet commented:
**Update 05/05/18**
- Styling fixes (per RytoEx review)     
- Support for Nvenc dynamic bitrate added (requires ffmpeg master with my commit https://git.ffmpeg.org/gitweb/ffmpeg.git/commit/155375123c1f2e41f97057f00e65bcbb20f7833a
which adds dynamic bitrate for nvenc encoder in ffmpeg)    
- Change of defaults after some testings (more aggressive when decreasing bitrate)    
- Improved algo for the bitrate increase phase : if congestion reappears, the bitrate is almost immediately returned to previous value (between 30-90 ms reaction time)
- bugfix when default bitrate is used (2500 kbs): since no bitrate is saved in streamEncoder.json , if one stopped a stream during a congestion and restarted later the initial bitrate was not reset to initial value (because no value was found in json).

RytoEX commented on UI/window-basic-status-bar.hpp:
```diff
@@ -6,6 +6,17 @@
+	BITRATE_EQUAL_INITIAL_BITRATE,
+	BITRATE_SWITCHING_DOWN,
+	BITRATE_SWITCHING_STATIONARY,//means bitrate is lower than initial bitrate so that
+								// the bitrate switching mechanism is still in place
```
These comments seem excessively indented?

pkviet replied:
you probably reviewed from github ? github display seems faulty; in VS2017 ide and notepad++ the  lines are aligned

RytoEX replied:
I always review from GitHub, initially.  In Notepad++, with a tab-size of 8 (this project's setting), it looks just like GitHub's representation.
![image](https://user-images.githubusercontent.com/624931/39700202-601769f0-51ca-11e8-9c1a-46d3ade18b3c.png)

In VS2017:
![image](https://user-images.githubusercontent.com/624931/39700326-c32cb05e-51ca-11e8-8928-98493ee6543a.png)

pkviet replied:
oh i guess i have tabs at 4 then ... shit, lemme fix that

RytoEX commented on UI/window-basic-status-bar.hpp:
```diff
@@ -6,6 +6,17 @@
+enum dynamicBitrateState {
+	BITRATE_EQUAL_INITIAL_BITRATE,
+	BITRATE_SWITCHING_DOWN,
+	BITRATE_SWITCHING_STATIONARY,//means bitrate is lower than initial bitrate so that
```
Nitpick:  Comment ends in punctuation, but does not seem to start as a sentence.

pkviet commented:
@RytoEX 
there are no issues if the avcodec dll is not up to date. The bitrate update block in obs-ffmpeg-nvenc sends an updated bitrate to nvenc which won't take it into account.
So no absolute need to put preprocessor directives but to be on the safe side I'll add some.  It requires avcodec 58.19.101 (bump from may 7 while my ffmpeg patch was merged on may 5)

pkviet commented:
**Update (07/18/18):**

- rebased to a single commit for better readibility.

Joefis-x20s commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -1065,9 +1119,166 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+
+	if (i_nal_hrd != 2 && stream->switch_variable_bitrate) {
+		/* Bitrate is adjusted downwards by 15% every second (on default
+		 * settings). The bitrate floor is arbitrarily set at 50 % of initial
```
I have build and tested this under my own conditions and have concluded that the bitrate floor should be user configurable. (It's the only problem I had with this) In some conditions for me the bitrate can vary more than 50%. Personally I would like to set my bitrate at 5000kb/s However, under heavy congestion I may only be able to achieve 1000kb/s even only 250kb/s at the worst times. I know the quality will be bad however I still prefer that than dropping frames. Other reasons is low end streams with an initial bitrate of 1000kb/s will only drop to 500kb/s under some conditions available upload may only be 200kb/s

I suggest making this user configurable as it can be very important in some cases.

pkviet replied:
I have no objection to letting the user set the floor bitrate.
50 % should be enough in most cases though because the resolution is not changed so lowering the bitrate even more is kind of violent. But I admit that some prefer to lower quality all the way rather than dropping frames.
Did you test several encoders? Qsv behaves better than the others.

Joefis-x20s replied:
Yeah I get you with not being able to change resolution. Just the odd times my upload does fall under 50% the bitrate floor it would be nice just to show something that looks a little bad as it's better than nothing by a lot. When I'm streaming to twitch at 6000kb/s bitrate the video can still look good at 1500kb/s for some less complex parts as that's usually my go to bit-rate when my upload is bad at rare points.

The main point I see from dynamic bitrate is just the goal of not dropping frames but still pushing the best quality the upload can handle. The bitrate floor can sometimes be a limiting factor for this and even if it falls below 50% I'm sure people would prefer the lower quality than dropping frames (Is there any benefit to dropping frames?) Also I couldn't get to test other encoders as I don't have QSV.

pkviet replied:
I'll add the floor setting next week along with other things

pkviet replied:
I'll add the floor setting next week along with other things

Joefis-x20s replied:
Thanks :) This PR was the only reason I went to find out how to build OBS since for me and others it can be such an important feature.

pkviet replied:
bitrate floor setting added.

pkviet commented:
**Update (08/02/18):**

- Bitrate floor setting added (sets the minimum bitrate; defaults to half the initial bitrate; some might want to have more aggressive settings.)    

- Defaults changed due to Twitch requirements for 30 sec between congestion detection and earliest bitrate increase : 
     - polling time between increases: every 30 sec ;
     - 35 % increase rate .

maxiannunziata commented:
im waiting this feaure guys!

pkviet commented:
**Update 10/08/2018**
Rebased to git head.
PR split in several commits.
Minor fixes.

pkviet commented:
**Update 10/08/2018**
Rebased to git head.
PR split in several commits.
Minor fixes (removal of a warning, function in obs_output_info moved to end of struct to avoid API break)

pkviet commented:
Update 11/08/2018
Rebased to git head.
PR split in several commits.
Minor fixes (removal of a warning, function in obs_output_info moved to end of struct to avoid API break)

ethaniel commented:
I need this feature as well!

TorreyBetts commented:
Whats the current status of this PR and getting it merged? There are quite a few IRL streamers that could benefit from this once it's merged.

Krutonium commented:
Any Chance you could provide a build of this? I managed to build it myself, but I need the browser source and TBH that went over my head at that point.

Krutonium commented:
I've been using this for about 2 weeks now, and frankly it's amazing. It's save me literally hundreds of thousands of dropped frames.

pkviet commented:
@Krutonium great to hear that. Don't forget to upvote https://ideas.obsproject.com/posts/349/dynamic-bitrate-support

Krutonium commented:
Does building with Dynamic Bitrate on nvenc require any extra steps? @pkviet

pkviet commented:
@Krutonium nope i've updated the PR to latest nvenc, both jim-nvenc and ffmpeg-nvenc (jim-nvenc is Jim's implementation which is more efficient) and dynamic bitrate works on both

Krutonium commented:
Awesome, so just build as normal then? *Wanders off to get started*

pkviet commented:
> Awesome, so just build as normal then? _Wanders off to get started_

right

dodgepong commented:
I think my only additional thought on this PR is that the arrow could use some better styling, and perhaps a tooltip on hover to describe what it is doing right now (if that's possible). Would it be possible to make the arrow green when increasing, and gray while even? Also, it looks like the arrow clips on the right when "even" so that the point isn't visible.

(This is based on the screenshots in the OP -- if this isn't the case anymore, then feel free to ignore.)

pkviet commented:
**update 02/21/19**
Per Dodgepong suggestions:
- changed colour of icons on the status bar;
- fixed the incomplete arrow;
- added localized tooltips.

dodgepong commented on UI/data/locale/en-US.ini:
```diff
@@ -424,6 +424,9 @@ Basic.StatusBar.Delay="Delay (%1 sec)"
 Basic.StatusBar.DelayStartingStoppingIn="Delay (stopping in %1 sec, starting in %2 sec)"
+Basic.StatusBar.DynamicBitrate.Up="Network congestion clearing.\nIncreasing bitrate."
+Basic.StatusBar.DynamicBitrate.Stationary="Network congestion still ongoing.\nBitrate stationary."
+Basic.StatusBar.DynamicBitrate.Down="Network congestion.\nDecreasing bitrate."
```
Maybe `Experiencing network congestion.\nDecreasing bitrate.`? I think that might read better.

pkviet replied:
ok will do

dodgepong commented on UI/data/locale/en-US.ini:
```diff
@@ -424,6 +424,9 @@ Basic.StatusBar.Delay="Delay (%1 sec)"
 Basic.StatusBar.DelayStoppingIn="Delay (stopping in %1 sec)"
 Basic.StatusBar.DelayStartingStoppingIn="Delay (stopping in %1 sec, starting in %2 sec)"
+Basic.StatusBar.DynamicBitrate.Up="Network congestion clearing.\nIncreasing bitrate."
+Basic.StatusBar.DynamicBitrate.Stationary="Network congestion still ongoing.\nBitrate stationary."
```
If this is the first time the user checks the tooltip, the "still" might throw them off. What about `Network congestion ongoing.\nMaintaining reduced bitrate.`?

pkviet replied:
will do

dodgepong commented:
...I'm an idiot, I didn't realize the "stationary" arrow only appeared during congestion. In that case I think it should be either red or yellow, not gray. Sorry, I know I'm being really nitpicky here ><

OBSAnon222 commented:
This looks great but how does one download this for use?

pkviet commented:
> This looks great but how does one download this for use?

This is a pull request, meaning a request for this code to be merged into the project. If you want to use these features, you'll have to compile binaries yourself (see the wiki for instructions) or wait until it's merged (or not).

hedgehog90 commented:
I found using the default NVENC encoder rather than the one marked **(new)**, didn't work as desired with dynamic variable bitrate.
It would occasionally flit between the lowered bitrate and the bitrate specified in the encoder settings (CBR), causing frame droppage during congestion.
However, selecting the **(new)** NVENC encoder fixed this issue.

hedgehog90 commented:
> This looks great but how does one download this for use?

I just compiled it for 64 bit windows: https://drive.google.com/file/d/17tO5KRQaSUFxcmlUtYVdagbcXazrc3VU/view

pkviet commented:
> I found using the default NVENC encoder rather than the one marked **(new)** didn't work as desired with dynamic variable bitrate.
> It would constantly flit between the lowered bitrate and the normal high bitrate specified in the encoder settings (CBR), causing frame droppage during congestion.
> However, selecting the **(new)** NVENC encoder fixed this issue.
> I'm curious to know if anyone else has experienced this.

Firstly, it's true that the new nvenc is better than the other encoders as to dynamic bitrate.
For the other encoders (x264, old nvenc, amd amf), you need to tighten the params:
- set for instance 100 ms for Decrease polling time instead of the 300 default
- increase the 'Bitrate decrease rate' (ex: 25% instead of 15 %).
In that way with CBR, old nvenc I completely avoid dropped frames.

pkviet commented:
**update**
- changed the default settings to decrease bitrate more aggressively ; will play better with old nvenc and x264
(see discussion with @hedgehog90 )

Ciommi commented:
If some pro can compile every new Version of Obs with this fantastic feature would be fantastic for nabs people like me 😋 thx anyway and amazing work

pkviet commented:
**update 03/30/19**
- major styling fixes ==> strict limit to 80 chars/line, shortening of var/function names, fix usage of CamelCase vs snake_case, 
- tried to improve readibility by splitting large functions (less 'squishy' code hopefully)
- added defaults to rtmp output instead of setting them in window-basic-main for Simple Output.
- removed long comments (wall of text).
- changed defaults (the 30 sec for increasing the bitrate is too long; this would pin the bitrate to lower values longer).
- mean congestion is computed now as a moving average , so I dropped all congestion arrays, which simplifies the algo.

For the sake of comparison I've left in the code an implementation of R1CH algo (adjust_bitrate-r1ch) written by Jim; it will have to be removed eventually.

Here's a comparison with R1CH original algo and the PR algo:
- in general, R1CH algo has more dropped frames (with jim-nvenc: x1.5, ffmpeg-nvenc: x2, x264: x2-x3) for a similar average bitrate.
- R1CH algo has a floor (min bitrate) set quite low at 100 kbs; for such a value, it has in general an average bitrate which is higher than the PR algo. But if one sets the floor for the PR algo to more realistic values, the average bitrate is similar.

For details with screen captures, check: https://docdro.id/Z5lwYX8

Krutonium commented:
This hasn't been touched in a month, could this be re-reviewed please? I have been using this build of OBS for a while now and I'd love to be able to switch back to mainline.

dodgepong commented:
It hasn't been touched because work has been more on getting 23.2 out the door. That said, Jim has been doing quite a bit of testing and tweaking with the algorithm internally lately. No need to be impatient.

Krutonium commented:
@pkviet I just found a crash that appears to be specific to this version of OBS - Specifically, when using QSV on an Intel HD 3000 (i7 2630QM) Laptop. This crash does not occur on the current versions of OBS.
[Crash.txt](https://github.com/obsproject/obs-studio/files/3317382/Crash.txt)

pkviet commented:
closing as Jim is implementing it himself

jp9000 commented:
For clarification I'm not implementing it myself, I just modified the algorithm.  I'll be contributing to this PR and adding some updates.  I'll also update it for clang-format.

pkviet commented:
feature coded by Jim with improvements has been merged so closing this.
